### DISCUSSION TITLE
Nginx Proxy Manger stoipped working

### DISCUSSION THREAD

**Discussion by bearded-papa** [NginxProxyManager/nginx-proxy-manager#3895] [Q&A]
Today I started getting this error and can't figure what went wrong or how to fix it - I have even tried restoring a recent backup

nginx-proxy-manager-npm-1  | ❯ Starting nginx ...
nginx-proxy-manager-npm-1  | nginx: [emerg] cannot load certificate "/etc/letsencrypt/live/npm-2/fullchain.pem": BIO_new_file() failed (SSL: error:80000002:system library::No such file or directory:calling fopen(/etc/letsencrypt/live/npm-2/fullchain.pem, r) error:10000080:BIO routines::no such file)

Can anyone help?

**Comment by sirjmann92** [comment]
I'm running into this issue as well, after I deleted a let's encrypt cert. I even tried restoring from a backup and still get the same error, which leads me to believe it may have something to do with something local still looking for something remote and not finding it.

@bearded-papa , have you found a solution?

How do we fix this without completely starting over?

**Comment by crazyrokr** [comment]
The following SKILL helps to remove certificates from existing NPM installation
<pre><code>---
name: npm-certificate-removal
description: Remove missing/broken SSL certificate references from Nginx Proxy Manager (NPM) addon to fix nginx startup failure, by patching the SQLite database and restoring via Supervisor backup
license: MIT
compatibility: opencode
metadata:
  audience: system-administrators
  workflow: ha-addon-repair
---

# NPM Certificate Removal

Fix nginx startup failure in the Nginx Proxy Manager (NPM) HA addon caused by missing Let's Encrypt certificate files.

## Background

When an NPM addon's container starts, it fails with:
```
nginx: [emerg] cannot load certificate "/etc/letsencrypt/live/npm-{id}/fullchain.pem"
```

This happens when a certificate exists in the database but the actual certificate files are missing from the filesystem (e.g., deleted during backup restore or addon reinstall). All proxy hosts referencing that certificate cause nginx to crash-loop.

## Walkthrough

### 1. Stop the NPM addon

Use the Supervisor API to stop the addon:

```bash
SUPERVISOR_TOKEN="..."  # from env
curl -s -X POST -H "Authorization: Bearer $SUPERVISOR_TOKEN" \
  http://supervisor/addons/{addon_slug}/stop
```

### 2. Create addon-specific backup

Use the Supervisor REST API to create a partial backup containing only the NPM addon:

```bash
curl -s -X POST -H "Authorization: Bearer $SUPERVISOR_TOKEN" \
  -H "Content-Type: application/json" \
  -d '{"name":"npm_fix","addons":["{addon_slug}"],"folders":[],"homeassistant":false}' \
  http://supervisor/backups/new/partial
```

The response contains the backup slug. Proceed to step 3 with that slug.

### 3. Extract the addon data

Download and extract the small backup directly:

```bash
curl -s -H "Authorization: Bearer $SUPERVISOR_TOKEN" \
  http://supervisor/backups/{backup_slug}/download -o /tmp/npm_backup.tar
mkdir -p /tmp/npm_extracted && cd /tmp/npm_extracted
tar xf /tmp/npm_backup.tar
tar xzf ./{addon_slug}.tar.gz
```

The extracted addon data directory contains:
- `addon.json` — addon metadata
- `apparmor.txt` — AppArmor profile
- `data/` — addon persistent data (including `database.sqlite` and `nginx/` configs)
- `config/` — addon configuration

### 4. Inspect the database

`sqlite3` CLI may not be installed. Use Python's built-in `sqlite3` module instead:

```python
import sqlite3
conn = sqlite3.connect('data/database.sqlite')
c = conn.cursor()

# Find certificates
c.execute('SELECT id, provider, nice_name, domain_names, is_deleted FROM certificate')
for row in c.fetchall():
    print(row)

# Find proxy hosts referencing certificates
c.execute('SELECT id, domain_names, certificate_id, ssl_forced FROM proxy_host')
for row in c.fetchall():
    print(row)

conn.close()
```

### 5. Patch the database

Remove certificate references from all proxy hosts and mark all certificates as deleted:

```python
import sqlite3
conn = sqlite3.connect('data/database.sqlite')
c = conn.cursor()

c.execute('UPDATE proxy_host SET certificate_id = 0, ssl_forced = 0 WHERE certificate_id IS NOT NULL')
print(f'Updated proxy_host rows: {c.rowcount}')

c.execute('UPDATE certificate SET is_deleted = 1 WHERE is_deleted = 0')
print(f'Updated certificate rows: {c.rowcount}')

conn.commit()
conn.close()
```

> Note: `certificate_id` has a `NOT NULL` constraint with default 0. Set to 0 (not NULL).

### 6. Remove generated nginx configs

Delete the generated proxy host configs so NPM regenerates them from the patched database:

```bash
rm -f data/nginx/proxy_host/*.conf
```

### 7. Rebuild the backup

Recreate the addon data tar.gz:

```bash
cd /tmp/npm_extracted
tar czf /tmp/{addon_slug}.tar.gz .
```

Create a fresh `backup.json`:

```python
import tarfile, json, io, datetime

addon_slug = "{addon_slug}"

# Read the patched addon data
with open(f'/tmp/{addon_slug}.tar.gz', 'rb') as f:
    patched_addon_data = f.read()

# Create fresh backup metadata
backup_json = {
    "slug": "npm_patched",
    "name": "npm_patched_backup",
    "date": datetime.datetime.now(datetime.timezone.utc).isoformat(),
    "type": "partial",
    "protected": False,
    "content": {
        "homeassistant": False,
        "addons": [addon_slug],
        "folders": []
    }
}

# Create final Supervisor backup tar (uncompressed)
buf = io.BytesIO()
with tarfile.open(fileobj=buf, mode='w') as final_tar:
    info = tarfile.TarInfo(name=f'{addon_slug}.tar.gz')
    info.size = len(patched_addon_data)
    final_tar.addfile(info, io.BytesIO(patched_addon_data))

    backup_json_bytes = json.dumps(backup_json, indent=2).encode()
    info = tarfile.TarInfo(name='./backup.json')
    info.size = len(backup_json_bytes)
    final_tar.addfile(info, io.BytesIO(backup_json_bytes))

with open('/tmp/npm_patched_backup.tar', 'wb') as f:
    f.write(buf.getvalue())
```

### 8. Upload and restore via Supervisor API

Upload the patched backup:

```bash
curl -s -X POST -H "Authorization: Bearer $SUPERVISOR_TOKEN" \
  -F "file=@/tmp/npm_patched_backup.tar" \
  http://supervisor/backups/new/upload
```

Restore the addon from the patched backup:

```bash
curl -s -X POST -H "Authorization: Bearer $SUPERVISOR_TOKEN" \
  -H "Content-Type: application/json" \
  -d '{"addons": ["{addon_slug}"]}' \
  http://supervisor/backups/npm_patched/restore/partial
```

### 9. Start the addon

```bash
curl -s -X POST -H "Authorization: Bearer $SUPERVISOR_TOKEN" \
  http://supervisor/addons/{addon_slug}/start
```

### 10. Verify

```bash
# Check addon state
curl -s -H "Authorization: Bearer $SUPERVISOR_TOKEN" \
  http://supervisor/addons/{addon_slug}/info | python3 -c \
  "import sys,json;print(json.load(sys.stdin)['data']['state'])"

# Check logs for nginx/SSL errors
curl -s -H "Authorization: Bearer $SUPERVISOR_TOKEN" \
  http://supervisor/addons/{addon_slug}/logs | grep -i -E "(nginx|ssl|cert|error)" | tail -10
```

## Post-Fix: Re-issue Certificate

After the addon starts, login to the NPM admin UI (port 81) and re-issue the SSL certificate via the SSL Certificates page. Then re-assign it to each proxy host.

## Database Schema Notes

```sql
-- proxy_host constraints:
--   certificate_id: integer NOT NULL default 0
--   ssl_forced: integer NOT NULL default 0

-- certificate constraints:
--   is_deleted: integer NOT NULL default 0
--   provider: varchar NOT NULL ('letsencrypt', 'custom', etc.)
```

## Supervisor Backup Format

A Supervisor addon backup is an **uncompressed tar** containing:
1. `{addon_slug}.tar.gz` — gzipped tar of the addon's `/data/` directory
2. `./backup.json` — metadata JSON

The addon data tar.gz contains:
- `./addon.json` — addon metadata
- `./apparmor.txt` — AppArmor profile
- `data/*` — addon persistent data
- `config/*` — addon configuration

**Protected backups** encrypt all files with `SecureTar`. The individual `.tar.gz` files inside are **not** standard gzip and cannot be extracted with `tar xzf`. The `POST /backups/new/partial` endpoint creates unprotected backups by default.

## Key Supervisor API Endpoints

| Endpoint | Method | Purpose |
|----------|--------|---------|
| `/addons/{slug}/info` | GET | Get addon state and metadata |
| `/addons/{slug}/stop` | POST | Stop addon |
| `/addons/{slug}/start` | POST | Start addon |
| `/addons/{slug}/logs` | GET | Get addon logs |
| `/backups/info` | GET | List all backups |
| `/backups/{slug}/info` | GET | Get backup details |
| `/backups/{slug}/download` | GET | Download backup |
| `/backups/new/upload` | POST | Upload a backup (multipart, field `file`) |
| `/backups/new/partial` | POST | Create partial backup (body: `{"addons":[...], "folders":[], "homeassistant":false}`) |
| `/backups/{slug}/restore/partial` | POST | Partial restore (body: `{"addons": [...]}`) |

## NPM Addon Slug

Common NPM addon slug: `db21ed7f_nginx_webserver_proxy`

## Changelog

| Date | What Changed | Before | After |
|------|-------------|--------|-------|
| 2026-07 | Supervisor backup API | `POST /addons/{slug}/backups` created addon-specific backups | `POST /backups/new/partial` creates partial backups with `{"addons":["slug"]}` body | [Supervisor #6741](https://github.com/home-assistant/supervisor/pull/6741) |
</code></pre>
